### PR TITLE
Retain backslashes in compiled SQL.

### PR DESCRIPTION
**Pull request by Ekrekr**
Fixes https://github.com/dataform-co/dataform-web-tracking/issues/200

If you write "\\\\" in the SQL part of SQLX, you'd it to act as "\\\\", rather than the JS interpretation of just "\\".

To solve this a new rule was added to the lexer. In addition to this, lexer tests were added, and an sqlx file to test.

Some minor whitespace changes made by prettier, my bad.

### CODE REVIEW

File: sqlx/lexer.ts [hunk 1]
```diff
@@ -309,9 +315,10 @@ function buildSqlxLexer(): { [x: string]: moo.Rules } {
     match: "${",
     push: LEXER_STATE_NAMES.JS_BLOCK
   };
-  sqlLexer[SQL_LEXER_TOKEN_NAMES.ESCAPED_BACKTICK] = {
-    match: "\\`"
-  };
+  // sqlLexer[SQL_LEXER_TOKEN_NAMES.ESCAPED_BACKTICK] = {
```
Comment (BenBirt) [inline sqlx/lexer.ts]: rm these lines?

File: tests/api/api.spec.ts [hunk 2]
```diff
@@ -520,7 +520,7 @@ describe("@dataform/api", () => {
   });
```
Comment (BenBirt) [inline tests/api/api.spec.ts]: ditto, plz revert this file

File: tests/api/examples.spec.ts [hunk 3]
```diff
@@ -101,8 +101,8 @@ describe("examples", () => {
           cleanSql(
             `select * from (select current_timestamp() as ts)
            where ts > (select max(ts) from \`tada-analytics.${schemaWithSuffix(
-             "df_integration_test"
-           )}.example_is_incremental\`) or (select max(ts) from \`tada-analytics.${schemaWithSuffix(
+              "df_integration_test"
```
Comment (BenBirt) [inline tests/api/examples.spec.ts]: this seems... wrong. is Prettier actually doing this?

File: tests/api/examples.spec.ts [hunk 4]
```diff
@@ -394,6 +394,17 @@ describe("examples", () => {
...
           "select 'hi' as faked union all\nselect 'ben' as faked union all\nselect 'sup?' as faked"
         );
+
+        // Check double back ticks don't get converted to singular.
+        const exampleDoubleBackslash = graph.tables.find(
+          (t: dataform.ITable) =>
+            t.name === schemaWithSuffix("df_integration_test") + ".example_double_backslash"
+        );
+        expect(exampleDoubleBackslash).to.not.be.undefined;
+        console.log("Raw query:", exampleDoubleBackslash.query);
```
Comment (BenBirt) [inline tests/api/examples.spec.ts]: rm this plz

File: tests/api/examples.spec.ts [hunk 5]
```diff
@@ -394,6 +394,17 @@ describe("examples", () => {
         expect(testCaseFQ.expectedOutputQuery.trim()).equals(
           "select 'hi' as faked union all\nselect 'ben' as faked union all\nselect 'sup?' as faked"
         );
+
+        // Check double back ticks don't get converted to singular.
```
Comment (BenBirt) [inline tests/api/examples.spec.ts]: "backslashes" - please correct comment

File: tests/sqlx/lexer.spec.ts [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+describe("@dataform/sqlx", () => {
+
+  describe("lexer", () => {
+    it("sqlx_single_backslash_stays_as_double", () => {
+      // These look like groups of 2 backslashes, but are actually 1. If someone
+      // in SQLX SQL writes "/" in the regex, it should be interpreted as "/".
+      const testCase = "SELECT regexp_extract('01a_data_engine', '^(\\d{2}\\w)')";
+      const parsedTestCase = parseSqlx(testCase);
+      const expectedResult = "SELECT regexp_extract('01a_data_engine', '^(\\\\d{2}\\\\w)')";
+      expect(parsedTestCase.sql[0]).equals(expectedResult);
```
Comment (BenBirt) [inline tests/sqlx/lexer.spec.ts]:
```suggestion
      expect(parsedTestCase.sql).equals([expectedResult]);
```

File: tests/sqlx/lexer.spec.ts [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+
+describe("@dataform/sqlx", () => {
+
+  describe("lexer", () => {
+    it("sqlx_single_backslash_stays_as_double", () => {
+      // These look like groups of 2 backslashes, but are actually 1. If someone
+      // in SQLX SQL writes "/" in the regex, it should be interpreted as "/".
+      const testCase = "SELECT regexp_extract('01a_data_engine', '^(\\d{2}\\w)')";
+      const parsedTestCase = parseSqlx(testCase);
+      const expectedResult = "SELECT regexp_extract('01a_data_engine', '^(\\\\d{2}\\\\w)')";
```
Comment (BenBirt) [inline tests/sqlx/lexer.spec.ts]: can we just inline this?

File: tests/sqlx/lexer.spec.ts [hunk 8]
```diff
@@ -0,0 +1,23 @@
+import { constructSyntaxTree, parseSqlx, lexer } from "@dataform/sqlx/lexer";
+import { expect } from "chai";
+
+describe("@dataform/sqlx", () => {
+
+  describe("lexer", () => {
+    it("sqlx_single_backslash_stays_as_double", () => {
+      // These look like groups of 2 backslashes, but are actually 1. If someone
+      // in SQLX SQL writes "/" in the regex, it should be interpreted as "/".
```
Comment (BenBirt) [inline tests/sqlx/lexer.spec.ts]: this comment (and below) refers to forward-slashes, i think you want backslashes.

File: tests/sqlx/lexer.spec.ts [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+      const testCase = "SELECT regexp_extract('01a_data_engine', '^(\\d{2}\\w)')";
+      const parsedTestCase = parseSqlx(testCase);
+      const expectedResult = "SELECT regexp_extract('01a_data_engine', '^(\\\\d{2}\\\\w)')";
+      expect(parsedTestCase.sql[0]).equals(expectedResult);
+    });
+    it("sqlx_double_backslash_stays_as_double", () => {
```
Comment (BenBirt) [inline tests/sqlx/lexer.spec.ts]: similar comments for this testcase

File: tests/sqlx/lexer.spec.ts [hunk 10]
```diff
@@ -0,0 +1,23 @@
+import { constructSyntaxTree, parseSqlx, lexer } from "@dataform/sqlx/lexer";
+import { expect } from "chai";
+
+describe("@dataform/sqlx", () => {
+
+  describe("lexer", () => {
+    it("sqlx_single_backslash_stays_as_double", () => {
```
Comment (BenBirt) [inline tests/sqlx/lexer.spec.ts]: nit: this should really read as a phrase, "replaces single backslashes with double backslashes"

File: yarn.lock [hunk 11]
```diff
@@ -6716,6 +6716,11 @@ no-case@^2.2.0, no-case@^2.3.2:
   dependencies:
     lower-case "^1.1.1"
 
+node-bin-setup@^1.0.0:
```
Comment (BenBirt) [inline yarn.lock]: do we need these new NPMs? i'd be very surprised! (similar comment for `package.json`)

File: core/compilers.ts [hunk 12]
```diff
@@ -124,8 +124,8 @@ switch (sqlxConfig.type) {
   case "inline": {
```
[the remark this answers is not in the document]
Comment (Ekrekr) [reply core/compilers.ts]: Sure thing

File: sqlx/lexer.ts [hunk 13]
```diff
@@ -58,7 +59,7 @@ const INNER_SQL_BLOCK_LEXER_TOKEN_NAMES = {
   CAPTURE_EVERYTHING_ELSE: LEXER_STATE_NAMES.INNER_SQL_BLOCK + "_captureEverythingElse"
 };
 
-const lexer = moo.states(buildSqlxLexer());
+export const lexer = moo.states(buildSqlxLexer());
```
[the remark this answers is not in the document]
Comment (Ekrekr) [reply sqlx/lexer.ts]: 👍

File: examples/common_v2/definitions/example_double_backslash.sqlx [hunk 14]
```diff
@@ -0,0 +1,9 @@
+config {
+    type: "view"
+}
+
+select * from regexp_extract('01a_data_engine', '^(\\d{2}\\w)')
+
+js {
```
[the remark this answers is not in the document]
Comment (Ekrekr) [reply examples/common_v2/definitions/example_double_backslash.sqlx]: 🚎

File: sqlx/lexer.ts (hunk 1 above)
Comment (Ekrekr) [reply sqlx/lexer.ts]: oops

File: tests/api/api.spec.ts (hunk 2 above)
Comment (Ekrekr) [reply tests/api/api.spec.ts]: 👍

File: tests/api/examples.spec.ts (hunk 5 above)
Comment (Ekrekr) [reply tests/api/examples.spec.ts]: 👍

File: tests/api/examples.spec.ts (hunk 4 above)
Comment (Ekrekr) [reply tests/api/examples.spec.ts]: Ah wow embarrassing

File: tests/sqlx/lexer.spec.ts (hunk 6 above)
Comment (Ekrekr) [reply tests/sqlx/lexer.spec.ts]: Ah that's nice

File: tests/sqlx/lexer.spec.ts (hunk 7 above)
Comment (Ekrekr) [reply tests/sqlx/lexer.spec.ts]: Yes much nicer

File: tests/sqlx/lexer.spec.ts (hunk 9 above)
Comment (Ekrekr) [reply tests/sqlx/lexer.spec.ts]: 👍

File: tests/sqlx/lexer.spec.ts (hunk 8 above)
Comment (Ekrekr) [reply tests/sqlx/lexer.spec.ts]: 👍

**Comment by Ekrekr** [comment]
My bad there's so many issues, didn't notice many because the diff was large from incorrect files being added.

File: tests/api/examples.spec.ts [hunk 15]
```diff
@@ -394,6 +394,16 @@ describe("examples", () => {
...
           "select 'hi' as faked union all\nselect 'ben' as faked union all\nselect 'sup?' as faked"
         );
+
+        // Check double backslashes don't get converted to singular.
+        const exampleDoubleBackslash = graph.tables.find(
+          (t: dataform.ITable) =>
+            t.name === schemaWithSuffix("df_integration_test") + ".example_double_backslash"
+        );
+        expect(exampleDoubleBackslash).to.not.be.undefined;
+        expect(cleanSql(exampleDoubleBackslash.query)).equals(
```
Comment (BenBirt) [inline tests/api/examples.spec.ts]: you should add an expectation on the value of `preOperations` (or whatever the field is called)

File: tests/sqlx/lexer.spec.ts [hunk 16]
```diff
@@ -0,0 +1,22 @@
+import { constructSyntaxTree, parseSqlx } from "@dataform/sqlx/lexer";
+import { expect } from "chai";
+
+describe("@dataform/sqlx", () => {
+  describe("lexer", () => {
+    it("replaces_single_backslash_with_double_backslashes", () => {
```
Comment (BenBirt) [inline tests/sqlx/lexer.spec.ts]: no_need_for_all_the_underscores

Comment (Ekrekr) [reply tests/sqlx/lexer.spec.ts]: O_K

File: tests/api/examples.spec.ts (hunk 15 above)
Comment (Ekrekr) [reply tests/api/examples.spec.ts]: Good idea. Turns out the SQL change didn't cover it!

File: examples/common_v2/definitions/example_double_backslash.sqlx [hunk 17]
```diff
@@ -0,0 +1,9 @@
+config {
+    type: "view"
+}
+
+select * from regexp_extract('01a_data_engine', '^(\\d{2}\\w)')
```
[the remark this answers is not in the document]
Comment (Ekrekr) [reply examples/common_v2/definitions/example_double_backslash.sqlx]: Added

File: tests/api/examples.spec.ts (hunk 15 above)
Comment (Ekrekr) [reply tests/api/examples.spec.ts]: Broke the whole thing, added a separate state to the lexer to keep track of entering and leaving single and double quotes.

File: sqlx/lexer.ts [hunk 18]
```diff
@@ -4,7 +4,9 @@ const LEXER_STATE_NAMES = {
   SQL: "sql",
   JS_BLOCK: "jsBlock",
   JS_TEMPLATE_STRING: "jsTemplateString",
-  INNER_SQL_BLOCK: "innerSqlBlock"
+  INNER_SQL_BLOCK: "innerSqlBlock",
+  INNER_SINGLE_QUOTE: "innerSingleQuote",
```
Comment (BenBirt) [inline sqlx/lexer.ts]: can we call these two `SQL_SINGLE_QUOTE_STRING` and `SQL_DOUBLE_QUOTE_STRING`?

File: sqlx/lexer.ts [hunk 19]
```diff
@@ -54,15 +55,29 @@ const INNER_SQL_BLOCK_LEXER_TOKEN_NAMES = {
   DOUBLE_QUOTE_STRING: LEXER_STATE_NAMES.INNER_SQL_BLOCK + "_doubleQuoteString",
   START_JS_PLACEHOLDER: LEXER_STATE_NAMES.INNER_SQL_BLOCK + "_startJsPlaceholder",
   CLOSE_BLOCK: LEXER_STATE_NAMES.INNER_SQL_BLOCK + "_closeBlock",
-  ESCAPED_BACKTICK: LEXER_STATE_NAMES.INNER_SQL_BLOCK + "_escapedBacktick",
+  BACKSLASH: LEXER_STATE_NAMES.INNER_SQL_BLOCK + "_backslash",
   BACKTICK: LEXER_STATE_NAMES.INNER_SQL_BLOCK + "_backtick",
-  CAPTURE_EVERYTHING_ELSE: LEXER_STATE_NAMES.INNER_SQL_BLOCK + "_captureEverythingElse"
+  CAPTURE_EVERYTHING_ELSE: LEXER_STATE_NAMES.INNER_SQL_BLOCK + "_captureEverythingElse",
+  START_QUOTE_SINGLE: LEXER_STATE_NAMES.INNER_SQL_BLOCK + "_startQuoteSingle",
```
Comment (BenBirt) [inline sqlx/lexer.ts]: please put these two above `CAPTURE_EVERYTHING_ELSE` (as you do when actually configuring the lexer)

File: sqlx/lexer.ts [hunk 20]
```diff
@@ -54,15 +55,29 @@ const INNER_SQL_BLOCK_LEXER_TOKEN_NAMES = {
...
+  BACKSLASH: LEXER_STATE_NAMES.INNER_DOUBLE_QUOTE + "_backslash",
+  CLOSE_QUOTE: LEXER_STATE_NAMES.INNER_DOUBLE_QUOTE + "_closeQuoteDouble",
+  CAPTURE_EVERYTHING_ELSE: LEXER_STATE_NAMES.INNER_DOUBLE_QUOTE + "_captureEverythingElse"
 };
 
 const lexer = moo.states(buildSqlxLexer());
 
 export interface ISyntaxTreeNode {
-  contentType: "sql" | "js" | "jsPlaceholder" | "sqlStatementSeparator" | "sqlComment";
+  contentType: "sql" | "js" | "jsPlaceholder" | "sqlStatementSeparator" | "sqlComment" | "quote";
```
Comment (BenBirt) [inline sqlx/lexer.ts]: i don't think we should make these changes... this implies callers of `constructSyntaxTree` must now handle this case as a separate content type, and (at least for now), quotes should just be contained as part of the `sql` content type.

File: sqlx/lexer.ts [hunk 21]
```diff
@@ -275,7 +293,20 @@ function getValueMappings() {
 
   valueMappings[INNER_SQL_BLOCK_LEXER_TOKEN_NAMES.STATEMENT_SEPERATOR] = () => "";
   valueMappings[INNER_SQL_BLOCK_LEXER_TOKEN_NAMES.CLOSE_BLOCK] = () => "";
+  valueMappings[INNER_SQL_BLOCK_LEXER_TOKEN_NAMES.SINGLE_LINE_COMMENT] = (value: string) =>
```
Comment (BenBirt) [inline sqlx/lexer.ts]: i'm a bit confused by the changes for `..._LINE_COMMENT` valueMappings - what is this changing? if it's fixing something, can you indicate with some changed test cases (and/or move to a different PR)?

File: sqlx/lexer.ts [hunk 22]
```diff
@@ -20,7 +22,6 @@ const SQL_LEXER_TOKEN_NAMES = {
   SINGLE_QUOTE_STRING: LEXER_STATE_NAMES.SQL + "_singleQuoteString",
```
Comment (BenBirt) [inline sqlx/lexer.ts]: shouldn't we be removing these two token types now? same with inner_sql?

File: tests/sqlx/lexer.spec.ts [hunk 23]
```diff
@@ -3,20 +3,42 @@ import { expect } from "chai";
...
+    });
+    it("closing braces in inner sql quotes doesn't close block", () => {
+      const parsedSqlx = parseSqlx("pre_operations {SELECT '{}'}");
+      expect(parsedSqlx.preOperations).eql(["SELECT '{}'"]);
+    });
+    it("closing braces in inner sql quotes with backslashes after doesn't close block", () => {
+      const parsedSqlx = parseSqlx("pre_operations {SELECT '\\d{2}'}");
+      expect(parsedSqlx.preOperations).eql(["SELECT '\\\\d{2}'"]);
+    });
+    it("escaped quotes do not close single quoted string", () => {
```
Comment (BenBirt) [inline tests/sqlx/lexer.spec.ts]: i'm not super sure what this test case is asserting. how would we know if it's closing the string or not? i think you might want to add some "special" text inside a still-open string, e.g. `js { foo = "bar" }` and see that the whole text is returned.

in general, since the PR changes how we lex strings fairly significantly, i think it would be a good idea to add some test cases for string matching - asserting that escaped end-quotes don't end the string, but also handling escaped backslashes, then escaped backslashes next to an un-escaped end-quote, etc.

File: sqlx/lexer.ts (hunk 18 above)
Comment (Ekrekr) [reply sqlx/lexer.ts]: Sure

File: sqlx/lexer.ts (hunk 20 above)
Comment (Ekrekr) [reply sqlx/lexer.ts]: Ah, that does make good sense

File: sqlx/lexer.ts (hunk 21 above)
Comment (Ekrekr) [reply sqlx/lexer.ts]: Hmm, I think they're residual from other fixes. My bad, will take them out

File: sqlx/lexer.ts (hunk 22 above)
Comment (Ekrekr) [reply sqlx/lexer.ts]: Ah good point

File: sqlx/lexer.ts (hunk 19 above)
Comment (Ekrekr) [reply sqlx/lexer.ts]: 👍

File: tests/sqlx/lexer.spec.ts (hunk 23 above)
Comment (Ekrekr) [reply tests/sqlx/lexer.spec.ts]: You are right, special text breaks it

Comment (Ekrekr) [reply tests/sqlx/lexer.spec.ts]: Screenshot 2019-12-18 at 16 39 31

Comment (Ekrekr) [reply tests/sqlx/lexer.spec.ts]: Issue was I hadn't added string handling to outer SQL, just inner SQL blocks. Have added more thorough tests now to make sure.

File: examples/common_v2/definitions/example_double_backslash.sqlx [hunk 24]
```diff
@@ -0,0 +1,11 @@
...
+
+select * from regexp_extract('01a_data_engine', r'^(\\d{2}\\w)')
```
Comment (BenBirt) [inline examples/common_v2/definitions/example_double_backslash.sqlx]: `r'^(\d{2}\w)'`

File: tests/sqlx/lexer.spec.ts [hunk 25]
```diff
@@ -0,0 +1,71 @@
+import { constructSyntaxTree, parseSqlx } from "@dataform/sqlx/lexer";
+import { expect } from "chai";
+
+describe("@dataform/sqlx", () => {
+  describe("outer SQL lexing", () => {
+    it("backslashes are duplicated so that they act as written in the IDE", () => {
+      // If someone in SQLX SQL writes "\" in the regex, it should be
+      // interpreted as "\".
+      const pairs = [
```
Comment (BenBirt) [inline tests/sqlx/lexer.spec.ts]: `const tests: { in: string; expected: string} = [ { ... }, { ... } ]`

File: tests/sqlx/lexer.spec.ts [hunk 26]
```diff
@@ -0,0 +1,71 @@
...
+        ["select 'asd\\'123\"def'", "select 'asd\\\\'123\"def'"],
+        ["select * from regexp_extract('js {', \"\")", "select * from regexp_extract('js {', \"\")"]
+      ];
+      pairs.forEach(pair => {
+        expect(parseSqlx(pair[0]).sql).eql([pair[1]]);
+      });
+    });
+  });
+  describe("inner SQL lexing", () => {
+    it("backslashes are duplicated so that they act as written in the IDE", () => {
```
Comment (BenBirt) [inline tests/sqlx/lexer.spec.ts]: don't talk about the IDE in open source code

File: examples/common_v2/definitions/example_double_backslash.sqlx (hunk 24 above)
Comment (Ekrekr) [reply examples/common_v2/definitions/example_double_backslash.sqlx]: 👍

File: tests/sqlx/lexer.spec.ts (hunk 25 above)
Comment (Ekrekr) [reply tests/sqlx/lexer.spec.ts]: 👍

File: tests/sqlx/lexer.spec.ts (hunk 26 above)
Comment (Ekrekr) [reply tests/sqlx/lexer.spec.ts]: Good point

File: sqlx/lexer.ts [hunk 27]
```diff
@@ -402,22 +414,28 @@ function buildSqlxLexer(): { [x: string]: moo.Rules } {
...
 
   const innerSingleQuoteLexer: moo.Rules = {};
+  innerSingleQuoteLexer[SQL_SINGLE_QUOTE_STRING_LEXER_TOKEN_NAMES.ESCAPED_BACKSLASH] = "\\\\";
+  innerSingleQuoteLexer[SQL_SINGLE_QUOTE_STRING_LEXER_TOKEN_NAMES.ESCAPED_QUOTE] = "\\'";
   innerSingleQuoteLexer[SQL_SINGLE_QUOTE_STRING_LEXER_TOKEN_NAMES.CLOSE_QUOTE] = {
     match: "'",
     pop: 1
   };
-  innerSingleQuoteLexer[SQL_SINGLE_QUOTE_STRING_LEXER_TOKEN_NAMES.BACKSLASH] = "\\";
+  // innerSingleQuoteLexer[SQL_SINGLE_QUOTE_STRING_LEXER_TOKEN_NAMES.BACKSLASH] = "\\";
```
Comment (BenBirt) [inline sqlx/lexer.ts]: rm commented out line

File: sqlx/lexer.ts [hunk 28]
```diff
@@ -402,22 +414,28 @@ function buildSqlxLexer(): { [x: string]: moo.Rules } {
...
+  innerDoubleQuoteLexer[SQL_DOUBLE_QUOTE_STRING_LEXER_TOKEN_NAMES.ESCAPED_BACKSLASH] = "\\\\";
+  innerDoubleQuoteLexer[SQL_DOUBLE_QUOTE_STRING_LEXER_TOKEN_NAMES.ESCAPED_QUOTE] = {
+    match: '\\"'
+  };
   innerDoubleQuoteLexer[SQL_DOUBLE_QUOTE_STRING_LEXER_TOKEN_NAMES.CLOSE_QUOTE] = {
     match: '"',
     pop: 1
   };
-  innerDoubleQuoteLexer[SQL_DOUBLE_QUOTE_STRING_LEXER_TOKEN_NAMES.BACKSLASH] = "\\";
+  // innerDoubleQuoteLexer[SQL_DOUBLE_QUOTE_STRING_LEXER_TOKEN_NAMES.BACKSLASH] = "\\";
```
Comment (BenBirt) [inline sqlx/lexer.ts]: ditto

File: tests/sqlx/lexer.spec.ts [hunk 29]
```diff
@@ -3,69 +3,145 @@ import { expect } from "chai";
 
 describe("@dataform/sqlx", () => {
   describe("outer SQL lexing", () => {
-    it("backslashes are duplicated so that they act as written in the IDE", () => {
+    it("backslashes are duplicated, so that they act literally when interpreted by javascript", () => {
```
Comment (BenBirt) [inline tests/sqlx/lexer.spec.ts]: nit: `... when included in a JavaScript template string`

File: tests/sqlx/lexer.spec.ts [hunk 30]
```diff
@@ -3,69 +3,145 @@ import { expect } from "chai";
...
-        [
-          "pre_operations {select regexp_extract('01a_data_engine', '^(\\\\d{2}\\\\w)')}",
-          "select regexp_extract('01a_data_engine', '^(\\\\\\\\d{2}\\\\\\\\w)')"
-        ],
-        [
-          "pre_operations {select * from regexp_extract('\\\\', '')}",
-          "select * from regexp_extract('\\\\\\\\', '')"
-        ],
-        ["pre_operations {}", ""]
+    it("backslashes are duplicated, so that they act literally when interpreted by javascript", () => {
```
Comment (BenBirt) [inline tests/sqlx/lexer.spec.ts]: ditto

File: tests/sqlx/lexer.spec.ts [hunk 31]
```diff
@@ -3,69 +3,145 @@ import { expect } from "chai";
...
       ];
-      pairs.forEach(pair => {
-        expect(parseSqlx(pair[0]).postOperations).eql([pair[1]]);
+      tests.forEach(test => {
+        expect(parseSqlx(test.in).postOperations).eql([test.expected]);
       });
     });
   });
+  describe("syntax tree construction", () => {
+    it("strings don't affect the tree", () => {
```
Comment (BenBirt) [inline tests/sqlx/lexer.spec.ts]: nit: `SQL strings don't ...`

File: tests/sqlx/lexer.spec.ts [hunk 32]
```diff
@@ -3,69 +3,145 @@ import { expect } from "chai";
...
+        contentType: "sql",
+        contents: [
+          'select regexp("", ',
+          { contentType: "jsPlaceholder", contents: ['${ref("dab")}'] },
+          ")"
+        ]
+      };
+      expect(tree).eql(expected);
+    });
+    it("inline js blocks tokenized correctly", () => {
```
Comment (BenBirt) [inline tests/sqlx/lexer.spec.ts]: this is a bit of an epic test, seems like it might be testing too much (it was useful for debugging, but not necessarily something we want to maintain in a test). can we boil it down to something more specific + smaller?

Comment (Ekrekr) [reply tests/sqlx/lexer.spec.ts]: Good point, have migrated the important string bit to the test above.
